netaddr_test.go

```diff
@@ -1441,31 +1443,38 @@ func BenchmarkIPv4Contains(b *testing.B) {
 	}
 }
 
-func BenchmarkParseIPv4(b *testing.B) {
-	b.ReportAllocs()
-	for i := 0; i < b.N; i++ {
-		ParseIP("192.168.1.1")
-	}
+var parseBenchInputs = []struct {
+	name string
+	ip   string
+}{
+	{"v4", "192.168.1.1"},
+	{"v6", "fd7a:115c:a1e0:ab12:4843:cd96:626b:430b"},
+	{"v6_ellipsis", "fd7a:115c::626b:430b"},
+	{"v6_v4", "::ffff:192.168.140.255"},
+	{"v6_zone", "1:2::ffff:192.168.140.255%eth1"},
 }
 
-func BenchmarkParseIPv6(b *testing.B) {
-	b.ReportAllocs()
-	for i := 0; i < b.N; i++ {
-		ParseIP("fe80::1cc0:3e8c:119f:c2e1%ens18")
-	}
-}
-
-func BenchmarkStdParseIPv4(b *testing.B) {
-	b.ReportAllocs()
-	for i := 0; i < b.N; i++ {
-		net.ParseIP("192.168.1.1")
+func BenchmarkParseIP(b *testing.B) {
+	z := intern.Get("eth1") // Pin to not benchmark the intern package
+	_ = z
+	for _, test := range parseBenchInputs {
```

Comment:
Any chance you'd want to reset the allocations count once the intern pool is set up?

Reply:
No change. Each case runs as a sub-benchmark, which I think implicitly resets the counters.

I'm still not sure where the extra alloc comes from in the v6+zone path. Either I'm doing the interning wrong and intern.Get allocates every time, or it's some unavoidable allocation in the intern package (maybe boxing the string in an interface{}?).

Reply:
Ack, unfortunately I don't have any more ideas.

Reply:
https://github.com/go4org/intern/pull/10 if we want it

Reply:
With GetString, things are happier:

```
name                  old time/op    new time/op    delta
ParseIP/v6_zone-8        221ns ± 3%     164ns ± 1%   -26.04%  (p=0.008 n=5+5)
StdParseIP/v6_zone-8     132ns ± 2%     134ns ± 2%      ~     (p=0.056 n=5+5)

name                  old alloc/op   new alloc/op   delta
ParseIP/v6_zone-8        16.0B ± 0%      0.0B       -100.00%  (p=0.008 n=5+5)
StdParseIP/v6_zone-8     16.0B ± 0%     16.0B ± 0%      ~     (all equal)

name                  old allocs/op  new allocs/op  delta
ParseIP/v6_zone-8         1.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
StdParseIP/v6_zone-8      1.00 ± 0%      1.00 ± 0%      ~     (all equal)
```

Still slightly slower than stdlib on v6+zone, but only very slightly, and because we're doing justified extra work to get nice value types.

Reply:
It's a nicely separable change and needs to wait until intern is updated, so I'm going to defer this until the next PR. This PR is still a net improvement in every dimension without the intern.GetString change.